[Ready for Review]New Azure Machine Learning api version 2021-07-01

zhaomuzhi opened the pull request:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [x] new API version 
      - [ ] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  -  [x] Ensure to copy the existing version into new directory structure for first commit (including refactoring) and then push new changes including version updates in separate commits. This is required to review the changes efficiently.
  - Adding a new service

- [x] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [x] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [x] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

zhaomuzhi committed:
Add Identity to createWorkpace example

zhaomuzhi committed:
Merge branch 'master' of https://github.com/zhaomuzhi/azure-rest-api-specs

zhaomuzhi committed:
Merge remote-tracking branch 'upstream/master'

zhaomuzhi committed:
Merge remote-tracking branch 'upstream/master'

zhaomuzhi committed:
Merge remote-tracking branch 'upstream/master'

zhaomuzhi committed:
Merge remote-tracking branch 'upstream/master'

zhaomuzhi committed:
Merge remote-tracking branch 'upstream/master'

zhaomuzhi committed:
change for getTransitivePEUsages

zhaomuzhi committed:
change for getTransitivePEUsages

zhaomuzhi committed:
update

zhaomuzhi committed:
update

zhaomuzhi committed:
prettier code

zhaomuzhi committed:
prettier code

zhaomuzhi committed:
add another new endpoint

zhaomuzhi committed:
add another new endpoint

zhaomuzhi committed:
address comments

zhaomuzhi committed:
address comments

zhaomuzhi committed:
revert change on transitivePE and notebookKeys

zhaomuzhi committed:
Merge branch 'master' of https://github.com/zhaomuzhi/azure-rest-api-specs

zhaomuzhi committed:
revert change on transitivePE and notebookKeys

zhaomuzhi committed:
Merge remote-tracking branch 'upstream/master'

zhaomuzhi committed:
Merge remote-tracking branch 'upstream/master'

zhaomuzhi committed:
add new stable version 2021-01-01

zhaomuzhi committed:
add ListNotebookAccessToken

zhaomuzhi committed:
add primaryUserAssignedIdentity

zhaomuzhi committed:
fix missing properties and apis

zhaomuzhi committed:
fix one more missing property

zhaomuzhi committed:
Merge remote-tracking branch 'upstream/master'

VinayRSoni committed:
202 status for start and stop CI

zhaomuzhi committed:
Merge pull request #8 from zhaomuzhi/vison/202startstopCI

202 status for start and stop CI

zhaomuzhi committed:
Merge remote-tracking branch 'upstream/master'

SubaMoorthy committed:
make resync keys long runnin operation in version Jan2021

SubaMoorthy committed:
resync keys 204

SubaMoorthy committed:
update resync keys example

SubaMoorthy committed:
Dont throw 204 for resync

SubaMoorthy committed:
update resync keys

SubaMoorthy committed:
Formatting

SubaMoorthy committed:
revert Jan2021changes and add new version

SubaMoorthy committed:
start apr2021 version for azure ml rp

SubaMoorthy committed:
undo remove new line in jan2021

zhaomuzhi committed:
Merge branch 'master' of https://github.com/SubaMoorthy/sukaruna-aml-apr2021

zhaomuzhi committed:
use systemData definition in common

zhaomuzhi committed:
removed old files

zhaomuzhi committed:
Merge pull request #3 from SubaMoorthy/referenceCommonSystemData

Reference common system data

SubaMoorthy committed:
fix small swagger bugs

SubaMoorthy committed:
Merge pull request #4 from SubaMoorthy/fix-swagger-kpi-bugs

fix small swagger bugs

SubaMoorthy committed:
run prettier

SubaMoorthy committed:
Merge pull request #6 from SubaMoorthy/fix-swagger-kpi-bugs

run prettier

SubaMoorthy committed:
remove reference to sku

SubaMoorthy committed:
Merge pull request #7 from SubaMoorthy/fix-swagger-kpi-bugs

remove reference to sku

SubaMoorthy committed:
update readme

SubaMoorthy committed:
Merge pull request #8 from SubaMoorthy/fix-swagger-kpi-bugs

update readme

Commit:
minimum agentCount should be 0

Commit:
add createdOn and ModifiedOn to compute properties

Commit:
add LoadBalancerType to aks compute

Commit:
add isNotebookInstanceCompute to virtualMachine

Commit:
add leafDomainLabel to SslConfiguration

Commit:
add overwriteExistingDomain to SslConfiguration

Commit:
add synapseSpark as supported compute-type

Commit:
update patch compute response to 202

Commit:
remove location from example responses

Commit:
Revert "remove location from example responses"

This reverts commit 048bf2551513afae842a9c4fbb5861aeddc6c33a.

joshanator committed:
Merge pull request #9 from SubaMoorthy/jopet/mlc-swagger-fixes

MLC swagger fixes

Commit:
MachineLearningCompute_Update should have a 200 response

SubaMoorthy committed:
sys data to use local def due to naming conflicts.

SubaMoorthy committed:
Merge pull request #10 from SubaMoorthy/fix-swagger-kpi-bugs

sys data to use local def due to naming conflicts.

SubaMoorthy committed:
Fix enum casing

SubaMoorthy committed:
Merge pull request #11 from SubaMoorthy/fix-swagger-kpi-bugs

Fix enum casing

SubaMoorthy committed:
Fix operationId PR comment.

SubaMoorthy committed:
Merge pull request #12 from SubaMoorthy/fix-swagger-kpi-bugs

Fix operationId PR comment.

SubaMoorthy committed:
Change opId in apr version

SubaMoorthy committed:
Merge pull request #13 from SubaMoorthy/fix-swagger-kpi-bugs

Change opId in apr version

SubaMoorthy committed:
remove 20210401 for easy review

SubaMoorthy committed:
only change api version

SubaMoorthy committed:
merge changes for 20210401 version

Commit:
Add x-ms-discriminator-value for Synapse spark

Commit:
camelcase loadbalancer properties

Commit:
add createdOn and Modified on to examples

Commit:
address AvoidAnonymousTypes error for SynapseSpark

SubaMoorthy committed:
Minor fixes + new property

SubaMoorthy committed:
Remove subscription state.

SubaMoorthy committed:
Merge branch 'Azure:master' into master

SubaMoorthy committed:
Merge branch 'Azure:master' into master

zhaomuzhi committed:
Merge branch 'master' of https://github.com/Azure/azure-rest-api-specs

zhaomuzhi committed:
Merge branch 'master' of https://github.com/Azure/azure-rest-api-specs into newVersion20210701_locked

zhaomuzhi committed:
copy over from baseline api version 2021-03-01-preview

zhaomuzhi committed:
removed preview features from 2021-03-01-preview and added more apis and properties in 2021-07-01

zhaomuzhi committed:
remove more preview features and examples

zhaomuzhi committed:
delete more examples of preview features

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:
```diff
@@ -0,0 +1,5067 @@
+    },
+    "VirtualMachineSizeListResult": {
+      "properties": {
+        "value": {
```
Glad to see we get this fixed! Thank you!

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:
```diff
@@ -0,0 +1,5067 @@
+          "200": {
+            "description": "The operation was successful.",
+            "schema": {
+              "$ref": "#/definitions/UpdateWorkspaceQuotasResult"
```
To confirm, does this request support pagination? I see we defined `nextlink` in this model

zhaomuzhi replied:
I think so, I don't own the service of this api though, did you see any problem?

ArcturusZhang replied:
I am just confirming that we are not making a mistake on this

zhaomuzhi committed:
do some fixes

zhaomuzhi committed:
add a new enum value in DiagnoseResult

mentat9 commented:
>             "UDR": {

ARM recommends using words instead of abbreviations and acroynms. At minimum, you should update the description to include the expanded abbreviation.

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:2664 in 3e92abd. [](commit_id = 3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03, deletion_comment = False)

mentat9 commented:
>             "NSG": {

Prefer full name to abbreviations.

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:2671 in 3e92abd. [](commit_id = 3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03, deletion_comment = False)

mentat9 commented:
>           "properties": {

These property names all need to be lowerCamelCase and plural. To make them plural, the names need to indicate what they are (diagnoseResults). So more like networkSecurityGroupDiagnoseResults, dnsResolutionDiagnoseResults, resourceLockDiagnoseResults, etc.

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:2663 in 3e92abd. [](commit_id = 3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03, deletion_comment = False)

mentat9 commented:
>       "post": {

Add a description.

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:302 in 3e92abd. [](commit_id = 3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03, deletion_comment = False)

mentat9 commented:
>       "get": {

Add a description.

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:1795 in 3e92abd. [](commit_id = 3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03, deletion_comment = False)

mentat9 commented:
>         }

Objects returned via the ARM resource model (GET/PUT/DELETE/PATCH) should put their properties into property bag.

Wiki link: [https://armwiki.azurewebsites.net/api_contracts/guidelines/rpc.html#rpc013-resource-definition-must-define-a-property-bag](https://armwiki.azurewebsites.net/api_contracts/guidelines/rpc.html#rpc013-resource-definition-must-define-a-property-bag)

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:5076 in 3e92abd. [](commit_id = 3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03, deletion_comment = False)

mentat9 commented:
>         "storageHnsEnabled": {

It sounds like this should be a string enum? Also, consider making this property easier to understand: using words instead of hns abbreviation (I don't know what it is), and maybe choosing a slightly different approach, e.g. storageType { Default, HydratedNoobStorage } that is easier for customers to understand, and more extensible.

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:2161 in 3e92abd. [](commit_id = 3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03, deletion_comment = False)

mentat9 commented:
>       "properties": {

If any of these contain secrets they aren't allowed in GET response payloads and need to be marked with x-ms-secret.

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:3283 in 3e92abd. [](commit_id = 3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03, deletion_comment = False)

mentat9 commented:
>           "additionalProperties": {

Avoid additionalProperties in favor of specific properties wherever possible. Nonschematized objects don't get support for important ARM features like Azure Policy and ARG.

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:3316 in 3e92abd. [](commit_id = 3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03, deletion_comment = False)

mentat9 commented:
>         "nodes": {

The collection array property should be named value.

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:4048 in 3e92abd. [](commit_id = 3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03, deletion_comment = False)

mentat9 commented:
>       "properties": {

It looks like the nextLink property was removed. If this is intentional, it is breaking change. The GET operations that use this definition are marked with x-ms-pageable, so I'm guessing it's an oversight.

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:4047 in 3e92abd. [](commit_id = 3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03, deletion_comment = False)

zhaomuzhi commented:
> > ```
> >   "properties": {
> > ```
> 
> It looks like the nextLink property was removed. If this is intentional, it is breaking change. The GET operations that use this definition are marked with x-ms-pageable, so I'm guessing it's an oversight.
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:4047 in [3e92abd](https://github.com/Azure/azure-rest-api-specs/commit/3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03). [](commit_id = [3e92abd](https://github.com/Azure/azure-rest-api-specs/commit/3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03), deletion_comment = False)

The ComputeNodesInformation already has nextLink defined, so we don't need to have a dup definition in AmlComputeNodesResult again

zhaomuzhi commented:
> > ```
> >     "nodes": {
> > ```
> 
> The collection array property should be named value.
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:4048 in [3e92abd](https://github.com/Azure/azure-rest-api-specs/commit/3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03). [](commit_id = [3e92abd](https://github.com/Azure/azure-rest-api-specs/commit/3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03), deletion_comment = False)

changing "nodes" to "value" will be a breaking change for existing api, also because AmlComputeNodesInformation is not planned to be used in list api, so I think it's up to the service to define what's the property name of an array?

zhaomuzhi commented:
> > ```
> >       "additionalProperties": {
> > ```
> 
> Avoid additionalProperties in favor of specific properties wherever possible. Nonschematized objects don't get support for important ARM features like Azure Policy and ARG.
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:3316 in [3e92abd](https://github.com/Azure/azure-rest-api-specs/commit/3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03). [](commit_id = [3e92abd](https://github.com/Azure/azure-rest-api-specs/commit/3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03), deletion_comment = False)

I checked with the feature owner, they said "additionalProperties" is the only acceptable option for them so far. Is it okay to keep it this way?

zhaomuzhi commented:
> > ```
> >     "storageHnsEnabled": {
> > ```
> 
> It sounds like this should be a string enum? Also, consider making this property easier to understand: using words instead of hns abbreviation (I don't know what it is), and maybe choosing a slightly different approach, e.g. storageType { Default, HydratedNoobStorage } that is easier for customers to understand, and more extensible.
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:2161 in [3e92abd](https://github.com/Azure/azure-rest-api-specs/commit/3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03). [](commit_id = [3e92abd](https://github.com/Azure/azure-rest-api-specs/commit/3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03), deletion_comment = False)

Sorry, I updated the property type to be "boolean", and updated the description to explain the Hns, "If the storage associated with the workspace has hierarchical namespace(HNS) enabled."

zhaomuzhi commented:
> > ```
> >   "get": {
> > ```
> 
> Add a description.
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:1795 in [3e92abd](https://github.com/Azure/azure-rest-api-specs/commit/3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03). [](commit_id = [3e92abd](https://github.com/Azure/azure-rest-api-specs/commit/3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03), deletion_comment = False)

We have "summary" defined for this api, is that good enough?

zhaomuzhi commented:
> > ```
> >   "get": {
> > ```
> 
> Add a description.
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:1795 in [3e92abd](https://github.com/Azure/azure-rest-api-specs/commit/3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03). [](commit_id = [3e92abd](https://github.com/Azure/azure-rest-api-specs/commit/3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03), deletion_comment = False)

We have summary defined for this api, is that good enough?

zhaomuzhi commented:
> > ```
> >   "post": {
> > ```
> 
> Add a description.
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:302 in [3e92abd](https://github.com/Azure/azure-rest-api-specs/commit/3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03). [](commit_id = [3e92abd](https://github.com/Azure/azure-rest-api-specs/commit/3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03), deletion_comment = False)

We have "summary" defined by this api, is that good enough?

zhaomuzhi committed:
fix bug

zhaomuzhi commented:
> > ```
> >         "NSG": {
> > ```
> 
> Prefer full name to abbreviations.
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:2671 in [3e92abd](https://github.com/Azure/azure-rest-api-specs/commit/3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03). [](commit_id = [3e92abd](https://github.com/Azure/azure-rest-api-specs/commit/3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03), deletion_comment = False)

zhaomuzhi committed:
update example

RichardZhaoW committed:
Merge branch 'Azure:master' into newVersion20210701_locked

Commit:
add x-ms-secret to some fields of KubernetesProperties

zhaomuzhi committed:
Merge pull request #23 from RichardZhaoW/newVersion20210701_locked

add x-ms-secret to some fields of KubernetesProperties

zhaomuzhi committed:
fix example

zhaomuzhi committed:
Merge branch 'newVersion20210701_locked' of https://github.com/zhaomuzhi/azure-rest-api-specs into newVersion20210701_locked

zhaomuzhi committed:
fix build error

mentat9 commented:
>         "storageHnsEnabled": {

String enum as suggested is preferred to boolean for several reasons. Details here: [https://armwiki.azurewebsites.net/rp_onboarding/process/api_review_best_practices.html?q=boolean#common-issues-found-in-review](https://armwiki.azurewebsites.net/rp_onboarding/process/api_review_best_practices.html?q=boolean#common-issues-found-in-review). This isn't blocking, but you should consider changing this to a string enum, and possibly improving it from a 2-state enabled/disabled switch to a more descriptive type property, e.g. storageType.

---
In reply to: [881750982](https://github.com/Azure/azure-rest-api-specs/pull/15141#issuecomment-881750982) [](http://example.com/codeflow?ancestors=881750982,881750982)

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:2161 in 3e92abd. [](commit_id = 3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03, deletion_comment = False)

mentat9 commented:
>       "properties": {

Thanks for adding the tag. If you leave these properties in the GET payload (which they currently are), make sure your RP sets them empty (or null) when retrieving. If you need a way for clients to retrieve them, you can return them in the listKeys action response.

---
In reply to: [881754944](https://github.com/Azure/azure-rest-api-specs/pull/15141#issuecomment-881754944) [](http://example.com/codeflow?ancestors=881754944,881754944)

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:3283 in 3e92abd. [](commit_id = 3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03, deletion_comment = False)

mentat9 commented:
I left a couple more comments, but I need to hand this off to this week's oncall (who is @majastrz) to complete the review.

zhaomuzhi commented:
> > ```
> >     "storageHnsEnabled": {
> > ```
> 
> String enum as suggested is preferred to boolean for several reasons. Details here: https://armwiki.azurewebsites.net/rp_onboarding/process/api_review_best_practices.html?q=boolean#common-issues-found-in-review. This isn't blocking, but you should consider changing this to a string enum, and possibly improving it from a 2-state enabled/disabled switch to a more descriptive type property, e.g. storageType.
> 
> In reply to: [881750982](https://github.com/Azure/azure-rest-api-specs/pull/15141#issuecomment-881750982)
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:2161 in [3e92abd](https://github.com/Azure/azure-rest-api-specs/commit/3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03). [](commit_id = [3e92abd](https://github.com/Azure/azure-rest-api-specs/commit/3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03), deletion_comment = False)

Thanks for the suggestion. For this one, we are pretty sure this property will only have two values, true or false. Because it's not a blocking issue and we have the service change ready, I will leave it as it is

zhaomuzhi committed:
add one new diagnosis category

majastrz commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/examples/Compute/createOrUpdate/AKSCompute.json:
```diff
@@ -0,0 +1,53 @@
+        "resourceId": "/subscriptions/34adfa4f-cedf-4dc0-ba29-b6d1a69ab345/resourcegroups/testrg123/providers/Microsoft.ContainerService/managedClusters/compute123-56826-c9b00420020b2",
+        "computeType": "AKS",
+        "properties": {
+          "agentCount": 4
```
>agentCount [](http://example.com/codeflow?start=11&length=10)

Why is `agentCount` missing from the responses?

majastrz replied:
And why do you have `properties` nested twice?

zhaomuzhi replied:
Hi, Marcin, we didn't update this example, it's inherited from previous versions, I'm not sure why it's defined like that, do we have to update this?

majastrz replied:
Does the API behave like the example suggests or is there a bug in the example?

ragargms replied:
agent count should be in responses. and yes, properties are nested twice. It's how the api has been defined since beginning.

zhaomuzhi replied:
fixed the example

majastrz commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/examples/Compute/get/AmlCompute.json:
```diff
@@ -0,0 +1,54 @@
+  },
+  "responses": {
+    "200": {
+      "body": {
```
>body [](http://example.com/codeflow?start=7&length=4)

The GET response here is very different than in the `createOrUpdate` example responses. Is that just an oversight or intentional?

zhaomuzhi replied:
I think it's oversight but because it's owned by a partner team, I can't speak for them. But since this example is inherited from previous versions, do we have to change it in this PR?

zhaomuzhi replied:
fixed the example

majastrz commented:
Added some comments and/or questions. Please take a look.

zhaomuzhi committed:
fix swagger example

majastrz reviewed (approved):
Signed off from ARM side.

RichardZhaoW commented:
> > ```
> >   "properties": {
> > ```
> 
> Thanks for adding the tag. If you leave these properties in the GET payload (which they currently are), make sure your RP sets them empty (or null) when retrieving. If you need a way for clients to retrieve them, you can return them in the listKeys action response.
> 
> In reply to: [881754944](https://github.com/Azure/azure-rest-api-specs/pull/15141#issuecomment-881754944)
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:3283 in [3e92abd](https://github.com/Azure/azure-rest-api-specs/commit/3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03). [](commit_id = [3e92abd](https://github.com/Azure/azure-rest-api-specs/commit/3e92abdbfae8ebc5bf84fae6a59f94e17e4c0b03), deletion_comment = False)

The 2 properties marked as "x-ms-secret" are always null in GET payload, and so far no client need to read these 2 properties.

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:
```diff
@@ -0,0 +1,5104 @@
+            "$ref": "#/definitions/AmlComputeNodeInformation"
+          },
+          "description": "The collection of returned AmlCompute nodes details."
+        }
```
Missing `nextLink` here - the operation with this as return schema has the extension `x-ms-pageable`

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:
```diff
@@ -0,0 +1,5104 @@
+    "AmlComputeNodesInformation": {
+      "description": "Compute node information related to a AmlCompute.",
+      "allOf": [
+        {
+          "$ref": "#/definitions/ComputeNodesInformation"
+        },
```
Since we are removing this type, we also need to put `nextlink` in the properties of this model.

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2021-07-01/machineLearningServices.json:
```diff
@@ -0,0 +1,5104 @@
+          "$ref": "#/definitions/ComputeNodesInformation"
+        },
+        {
+          "$ref": "#/definitions/AmlComputeNodesResult"
```
The polymorphism is redundant here, please remove
```suggestion
```

zhaomuzhi committed:
address comments on ListNodes api
